File: src/exporter.py
"""報告匯出器 - 匯出 JSON 報告與 EDL 檔案"""

import json
from datetime import datetime
from pathlib import Path
from typing import Optional

from .editor import EditReport


class ReportExporter:
    """報告匯出器

    支援匯出 JSON 報告與 EDL 檔案。
    """

    @staticmethod
    def to_json(
        report: EditReport,
        output_path: Path,
        pretty: bool = True
    ) -> None:
        """
        匯出 JSON 報告

        Args:
            report: 編輯報告
            output_path: 輸出路徑
            pretty: 是否美化輸出
        """
        data = {
            "version": "1.0",
            "generated_at": datetime.now().isoformat(),
            "input": str(report.input_path),
            "output": str(report.output_path),
            "original_duration": report.original_duration,
            "edited_duration": report.edited_duration,
            "removed_duration": report.original_duration - report.edited_duration,
            "reduction_percent": (
                (report.original_duration - report.edited_duration)
                / report.original_duration * 100
                if report.original_duration > 0 else 0
            ),
            "edit_count": len(report.edits),
            "edits": [e.to_dict() for e in report.edits],
            "statistics": ReportExporter._compute_statistics(report)
        }

        output_path.parent.mkdir(parents=True, exist_ok=True)

        indent = 2 if pretty else None
        output_path.write_text(
            json.dumps(data, ensure_ascii=False, indent=indent),
            encoding="utf-8"
        )

    @staticmethod
    def _compute_statistics(report: EditReport) -> dict:
        """計算統計資訊"""
        stats = {
            "by_reason": {},
            "total_removed_duration": 0.0,
        }

        for edit in report.edits:
            reason = edit.reason
            duration = edit.original_end - edit.original_start

            if reason not in stats["by_reason"]:
                stats["by_reason"][reason] = {
                    "count": 0,
                    "duration": 0.0
                }

            stats["by_reason"][reason]["count"] += 1
            stats["by_reason"][reason]["duration"] += duration
            stats["total_removed_duration"] += duration

        return stats

    @staticmethod
    def to_edl(
        report: EditReport,
        output_path: Path,
        fps: float = 30.0,
        title: Optional[str] = None
    ) -> None:
        """
        匯出 EDL 檔案 (供 DAW 使用)

        EDL (Edit Decision List) 格式可被 DaVinci Resolve、
        Adobe Premiere、Pro Tools 等軟體讀取。

        Args:
            report: 編輯報告
            output_path: 輸出路徑
            fps: 影格率（用於時間碼轉換）
            title: EDL 標題
        """
        lines = []

        # EDL 標頭
        edl_title = title or report.input_path.stem
        lines.append(f"TITLE: {edl_title}")
        lines.append(f"FCM: NON-DROP FRAME")
        lines.append("")

        # 計算保留的區間
        keep_regions = ReportExporter._compute_keep_regions(report)

        # 生成 EDL 事件
        for i, (start, end) in enumerate(keep_regions, 1):
            event_num = f"{i:03d}"
            reel = "AX"  # Audio eXternal

            # 轉換時間碼
            src_in = ReportExporter._seconds_to_timecode(start, fps)
            src_out = ReportExporter._seconds_to_timecode(end, fps)

            # 計算記錄時間碼（累計時間）
            rec_start = sum(
                keep_regions[j][1] - keep_regions[j][0]
                for j in range(i - 1)
            )
            rec_end = rec_start + (end - start)

            rec_in = ReportExporter._seconds_to_timecode(rec_start, fps)
            rec_out = ReportExporter._seconds_to_timecode(rec_end, fps)

            # EDL 行格式
            line = f"{event_num}  {reel}       AA/V  C        {src_in} {src_out} {rec_in} {rec_out}"
            lines.append(line)

            # 檔案名稱註解
            lines.append(f"* FROM CLIP NAME: {report.input_path.name}")
            lines.append("")

        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text("\n".join(lines), encoding="utf-8")

    @staticmethod
    def _compute_keep_regions(report: EditReport) -> list:
        """計算保留的區間"""
        if not report.edits:
            return [(0.0, report.original_duration)]

        # 排序編輯
        sorted_edits = sorted(report.edits, key=lambda e: e.original_start)

        regions = []
        last_end = 0.0

        for edit in sorted_edits:
            if edit.original_start > last_end:
                regions.append((last_end, edit.original_start))
            last_end = edit.original_end

        # 最後一段
        if last_end < report.original_duration:
            regions.append((last_end, report.original_duration))

        return regions

    @staticmethod
    def _seconds_to_timecode(seconds: float, fps: float = 30.0) -> str:
        """
        將秒數轉換為時間碼格式 (HH:MM:SS:FF)

        Args:
            seconds: 秒數
            fps: 影格率

        Returns:
            時間碼字串
        """
        total_frames = int(seconds * fps)

        frames = total_frames % int(fps)
        total_seconds = total_frames // int(fps)

        secs = total_seconds % 60
        total_minutes = total_seconds // 60

        mins = total_minutes % 60
        hours = total_minutes // 60

        return f"{hours:02d}:{mins:02d}:{secs:02d}:{frames:02d}"

    @staticmethod
    def to_markers(
        report: EditReport,
        output_path: Path,
        format: str = "csv"
    ) -> None:
        """
        匯出標記檔案（可用於 Audacity、Reaper 等）

        Args:
            report: 編輯報告
            output_path: 輸出路徑
            format: 輸出格式 ("csv" 或 "txt")
        """
        output_path.parent.mkdir(parents=True, exist_ok=True)

        if format == "csv":
            lines = ["start,end,label,reason"]
            for edit in report.edits:
                # 處理文字中的逗號和換行
                text = edit.text.replace(",", ";").replace("\n", " ")
                lines.append(
                    f"{edit.original_start:.3f},{edit.original_end:.3f},"
                    f"\"{text}\",{edit.reason}"
                )
            output_path.write_text("\n".join(lines), encoding="utf-8")

        else:  # txt (Audacity label format)
            lines = []
            for edit in report.edits:
                text = edit.text.replace("\t", " ").replace("\n", " ")
                lines.append(
                    f"{edit.original_start:.6f}\t{edit.original_end:.6f}\t"
                    f"[{edit.reason}] {text}"
                )
            output_path.write_text("\n".join(lines), encoding="utf-8")

File: crates/reclip-core/src/exporter.rs
//! 報告匯出模組

use std::collections::HashMap;
use std::fs;
use std::io::Write;
use std::path::Path;

use chrono::Local;
use serde::Serialize;
use thiserror::Error;

use crate::types::EditReport;

/// 匯出錯誤
#[derive(Error, Debug)]
pub enum ExportError {
    #[error("IO 錯誤: {0}")]
    Io(#[from] std::io::Error),

    #[error("JSON 序列化錯誤: {0}")]
    Json(#[from] serde_json::Error),
}

/// 報告匯出器
pub struct Exporter;

impl Exporter {
    /// 匯出 JSON 報告
    pub fn to_json<P: AsRef<Path>>(
        report: &EditReport,
        output_path: P,
        pretty: bool,
    ) -> Result<(), ExportError> {
        let output_path = output_path.as_ref();

        // 確保目錄存在
        if let Some(parent) = output_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let data = JsonReport::from_report(report);

        let json = if pretty {
            serde_json::to_string_pretty(&data)?
        } else {
            serde_json::to_string(&data)?
        };

        fs::write(output_path, json)?;
        Ok(())
    }

    /// 匯出 EDL 檔案
    pub fn to_edl<P: AsRef<Path>>(
        report: &EditReport,
        output_path: P,
        fps: f64,
        title: Option<&str>,
    ) -> Result<(), ExportError> {
        let output_path = output_path.as_ref();

        if let Some(parent) = output_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let edl_title = title.unwrap_or_else(|| {
            Path::new(&report.input_path)
                .file_stem()
                .and_then(|s| s.to_str())
                .unwrap_or("Untitled")
        });

        let mut content = String::new();
        content.push_str(&format!("TITLE: {}\n", edl_title));
        content.push_str("FCM: NON-DROP FRAME\n\n");

        // 計算保留的區間
        let keep_regions = Self::compute_keep_regions(report);

        let mut rec_offset = 0.0;
        for (i, (start, end)) in keep_regions.iter().enumerate() {
            let event_num = format!("{:03}", i + 1);
            let reel = "AX";

            let src_in = Self::seconds_to_timecode(*start, fps);
            let src_out = Self::seconds_to_timecode(*end, fps);

            let duration = end - start;
            let rec_in = Self::seconds_to_timecode(rec_offset, fps);
            let rec_out = Self::seconds_to_timecode(rec_offset + duration, fps);

            content.push_str(&format!(
                "{}  {}       AA/V  C        {} {} {} {}\n",
                event_num, reel, src_in, src_out, rec_in, rec_out
            ));

            let input_name = Path::new(&report.input_path)
                .file_name()
                .and_then(|s| s.to_str())
                .unwrap_or("input");
            content.push_str(&format!("* FROM CLIP NAME: {}\n\n", input_name));

            rec_offset += duration;
        }

        fs::write(output_path, content)?;
        Ok(())
    }

    /// 匯出標記檔案
    pub fn to_markers<P: AsRef<Path>>(
        report: &EditReport,
        output_path: P,
        format: MarkerFormat,
    ) -> Result<(), ExportError> {
        let output_path = output_path.as_ref();

        if let Some(parent) = output_path.parent() {
            fs::create_dir_all(parent)?;
        }

        let content = match format {
            MarkerFormat::Csv => Self::format_markers_csv(report),
            MarkerFormat::Audacity => Self::format_markers_audacity(report),
        };

        fs::write(output_path, content)?;
        Ok(())
    }

    /// 計算保留的區間
    fn compute_keep_regions(report: &EditReport) -> Vec<(f64, f64)> {
        if report.edits.is_empty() {
            return vec![(0.0, report.original_duration)];
        }

        let mut sorted_edits = report.edits.clone();
        sorted_edits.sort_by(|a, b| a.original_start.partial_cmp(&b.original_start).unwrap());

        let mut regions = Vec::new();
        let mut last_end = 0.0;

        for edit in &sorted_edits {
            if edit.original_start > last_end {
                regions.push((last_end, edit.original_start));
            }
            last_end = edit.original_end;
        }

        if last_end < report.original_duration {
            regions.push((last_end, report.original_duration));
        }

        regions
    }

    /// 將秒數轉換為時間碼
    fn seconds_to_timecode(seconds: f64, fps: f64) -> String {
        let total_frames = (seconds * fps) as u32;
        let frames = total_frames % fps as u32;
        let total_seconds = total_frames / fps as u32;
        let secs = total_seconds % 60;
        let total_minutes = total_seconds / 60;
        let mins = total_minutes % 60;
        let hours = total_minutes / 60;

        format!("{:02}:{:02}:{:02}:{:02}", hours, mins, secs, frames)
    }

    /// 格式化為 CSV 標記
    fn format_markers_csv(report: &EditReport) -> String {
        let mut lines = vec!["start,end,label,reason".to_string()];

        for edit in &report.edits {
            let text = edit.text.replace(',', ";").replace('\n', " ");
            lines.push(format!(
                "{:.3},{:.3},\"{}\",{}",
                edit.original_start, edit.original_end, text, edit.reason
            ));
        }

        lines.join("\n")
    }

    /// 格式化為 Audacity 標記
    fn format_markers_audacity(report: &EditReport) -> String {
        let mut lines = Vec::new();

        for edit in &report.edits {
            let text = edit.text.replace('\t', " ").replace('\n', " ");
            lines.push(format!(
                "{:.6}\t{:.6}\t[{}] {}",
                edit.original_start, edit.original_end, edit.reason, text
            ));
        }

        lines.join("\n")
    }
}

/// 標記格式
pub enum MarkerFormat {
    /// CSV 格式
    Csv,
    /// Audacity 標籤格式
    Audacity,
}

/// JSON 報告結構
#[derive(Serialize)]
struct JsonReport {
    version: String,
    generated_at: String,
    input: String,
    output: String,
    original_duration: f64,
    edited_duration: f64,
    removed_duration: f64,
    reduction_percent: f64,
    edit_count: usize,
    edits: Vec<JsonEdit>,
    statistics: Statistics,
}

#[derive(Serialize)]
struct JsonEdit {
    original_start: f64,
    original_end: f64,
    reason: String,
    text: String,
}

#[derive(Serialize)]
struct Statistics {
    by_reason: HashMap<String, ReasonStats>,
    total_removed_duration: f64,
}

#[derive(Serialize)]
struct ReasonStats {
    count: u32,
    duration: f64,
}

impl JsonReport {
    fn from_report(report: &EditReport) -> Self {
        let removed_duration = report.original_duration - report.edited_duration;
        let reduction_percent = if report.original_duration > 0.0 {
            removed_duration / report.original_duration * 100.0
        } else {
            0.0
        };

        // 計算統計
        let mut by_reason: HashMap<String, ReasonStats> = HashMap::new();
        let mut total_removed = 0.0;

        for edit in &report.edits {
            let duration = edit.original_end - edit.original_start;
            total_removed += duration;

            let stats = by_reason.entry(edit.reason.clone()).or_insert(ReasonStats {
                count: 0,
                duration: 0.0,
            });
            stats.count += 1;
            stats.duration += duration;
        }

        Self {
            version: "1.0".to_string(),
            generated_at: Local::now().to_rfc3339(),
            input: report.input_path.clone(),
            output: report.output_path.clone(),
            original_duration: report.original_duration,
            edited_duration: report.edited_duration,
            removed_duration,
            reduction_percent,
            edit_count: report.edits.len(),
            edits: report
                .edits
                .iter()
                .map(|e| JsonEdit {
                    original_start: e.original_start,
                    original_end: e.original_end,
                    reason: e.reason.clone(),
                    text: e.text.clone(),
                })
                .collect(),
            statistics: Statistics {
                by_reason,
                total_removed_duration: total_removed,
            },
        }
    }
}
